Web Docs - Stepper Component (#2727)
hashicorp/design-system: website/docs/components/stepper/list/partials/guidelines/guidelines.md

Reviewer: The related section is missing from this page.

Author: Added.

## Usage

### When to use

> - In complex flows that are broken up into multiple steps to more easily segment content.

Author: Good suggestion will update

- To communicate progress in a mult-step flow; what has been completed, and what has yet to be completed.
- For progress that does not require a user to remain on a single page to complete.

### When not to use 

- For page level, linear progress that requires a user to remain on the page for, use the [Stepper Nav](/components/stepper/nav).
- Don't require specific progress tracking, use structured page content.

## Content

> The Stepper has a title, an optional description and a generic slot for content.

Author: Sure, ill make that update. Good point Jory.

### Title

> Titles are required and help users understand, from an overview perspective, what tasks need to be completed. Titles should be brief, and only a couple of words.

Reviewer 2: To further this suggestion:

> Titles provide an overview to help users understand what tasks need to be completed and their purpose within the flow.

Yes, the title is communicating and sectioning the tasks, but should also be clear enough that a user has an idea of what needs to be done.

### Description

Descriptions are optional and should be brief and provide additional context to the title. We recommend that all steps either have a description or none do.

!!! Do

![The stepper with descriptions enabled in every step.](/assets/components/stepper/list/stepper-list-description-do.png)

!!!

!!! Dont

![Steps one and four with enabled descriptions, while two and three have only a title.](/assets/components/stepper/list/stepper-list-description-dont.png)

!!!

## Progressive disclosure

We recommend that the contents of the in-progress step be displayed at a time. This narrows the focus on the specific task at hand and ensures that they are completed in the specified order. However, this isn’t always possible and sometimes context from the previous step allows the user to complete the next step.

> ![All steps only showing a title and description other than step 2. Step 2 shows a title, description and a button.](/assets/components/stepper/list/stepper-list-progressive-disclosure.png)

Author: How does this look?

> ![All steps only showing a title and description other than step 2. Step 2 shows a title, description and a button.](/assets/components/stepper/list/stepper-list-progressive-disclosure.png)

Reviewer: I don't see any difference.

> ![All steps only showing a title and description other than step 2. Step 2 shows a title, description and a button.](/assets/components/stepper/list/stepper-list-progressive-disclosure.png)

Author: Oops, just updated.

> ![All steps only showing a title and description other than step 2. Step 2 shows a title, description and a button.](/assets/components/stepper/list/stepper-list-progressive-disclosure.png)

Reviewer: This is what I'm seeing still which doesn't use a real world example: 
<img width="693" alt="Screenshot 2025-03-06 at 1 53 06 PM" src="https://github.com/user-attachments/assets/ff034a9b-9418-471c-ad49-8355c9c97b7c" />

Is that to be expected?

> ![All steps only showing a title and description other than step 2. Step 2 shows a title, description and a button.](/assets/components/stepper/list/stepper-list-progressive-disclosure.png)

Author: Updated, sorry I don't know how I missed this notification.

## Using processing status

> If a Step’s contents have submission capabilities, a processing status can help the user understand that their data is in the process of being submitted.

Reviewer 2: [suggestion] could also add something about a step kicking off a background process for this variant.

> ![A Stepper list with two steps. The first step the user is pressing a button, and the status of the first step is in a processing state.](/assets/components/stepper/list/stepper-list-processing-status.png)

Author: Small, will fix!

## Usage

### When to use

> - In onboarding, account creation or initial setups that are broken up into multiple steps to more easily segment content.

Reviewer 3: ```suggestion
- In onboarding, account creation, or initial setups that are broken up into multiple steps to more easily segment content.
```
[nitpick] Oxford comma

> - In onboarding, account creation, or initial setups that are broken up into multiple steps to more easily segment content.

Reviewer 4: ```suggestion
- In onboarding, account creation, or initial setups that are broken up into multiple steps.
```

Reviewer 4: the last part seems a bit redundant to me.

- To communicate what has and has not been completed in a multi-step flow.

> - For progress that does not require a user to remain on a single page to complete.

Reviewer 3: ```suggestion
- For progress that does not require a user to remain on the current page to complete.
```
[nitpick/suggestion] not sure if this is clearer/more accurate

### When not to use

- When the user needs to remain within the step to complete it, use the [Stepper Nav](/components/stepper/nav).
- For simple forms or when progress doesn't need to be tracked, use structured page content.

## Content

The Stepper List has a title, an optional description, and a generic slot for content.

### Title

Titles provide an overview to help users understand what tasks need to be completed and their purpose within the flow. Titles are required and should be brief.

### Description

> Descriptions are optional. They should provide additional context but be brief. We recommend that all steps either have a description or do not.

Reviewer 4: ```suggestion
Descriptions are optional. They can be used to provide additional brief context. If used, we recommend providing a description for all steps.
```

!!! Do

> Ensure description visibility is consistent.

Reviewer 4: ```suggestion
Ensure inclusion of descriptions is consistent.
```

Reviewer 4: There may be other words that are better but I think this is more accurate than "visibility" which to me implies that some descriptions could be invisible or temporarily hidden but still there.

![The stepper with descriptions enabled in every step.](/assets/components/stepper/list/stepper-list-description-do.png)

!!!

!!! Dont

> Don't mix and match description visibility.

Reviewer 4: ```suggestion
Don't mix and match inclusion of descriptions.
```

Reviewer 4: Same as above. There may be other ways to say this but I think using the word "visibility" has more potential for misunderstanding.

> ![Steps one and four with enabled descriptions, while two and three have only a title.](/assets/components/stepper/list/stepper-list-description-dont.png)

Reviewer 4: ```suggestion
![Steps one and four include descriptions, while steps two and three only have titles.](/assets/components/stepper/list/stepper-list-description-dont.png)
```

!!!

## Progressive disclosure

> We recommend showing only the contents of the **in-progress** step. This narrows the focus to the specific step and ensures steps are completed in order. However, this isn’t always possible, and sometimes context from the previous step allows the user to complete the next step. Use your best judgment.

Reviewer 4: ```suggestion
We recommend showing the contents of only the **in-progress** step while hiding the contents of other steps. This narrows the focus to only the current step and ensures steps are completed in order. However, this isn’t always feasible, and sometimes context from the previous step is needed for the user to complete the next step. Use your best judgment.
```

![All steps only showing a title and description other than step 2. Step 2 shows a title, description and a button.](/assets/components/stepper/list/stepper-list-progressive-disclosure.png)

## Using processing status

> If data can be submitted within a step, showing the status as `processing` provides the user with immediate feedback.

Reviewer 4: ```suggestion
Set the status as `processing` to provide the user with immediate feedback when data is submitted within a step.
```

> ![A Stepper list with two steps. The first step the user is pressing a button, and the status of the first step is in a processing state.](/assets/components/stepper/list/stepper-list-processing-status.png)

Reviewer 4: ```suggestion
![A Stepper list with two steps. In the first step, the user is pressing a button and the step is in a processing state.](/assets/components/stepper/list/stepper-list-processing-status.png)
```

## Usage

### When to use

- In onboarding, account creation, or initial setups that are broken up into multiple steps.
- To communicate what has and has not been completed in a multi-step flow.

> - For progress that does not require a user to remain on the current page to complete.

Reviewer 4: ```suggestion
- For progress that does not require the user to remain on the current page to complete.
```

### When not to use

- When the user needs to remain within the step to complete it, use the [Stepper Nav](/components/stepper/nav).
- For simple forms or when progress doesn't need to be tracked, use structured page content.

## Content

The Stepper List includes a title, an optional description, and a generic slot for content.

### Title

Titles provide an overview to help users understand what tasks need to be completed and their purpose within the flow. Titles are required and should be brief.

### Description

Descriptions are optional. They can be used to provide additional brief context. If used, we recommend providing a description for all steps.

!!! Do

Ensure inclusion of descriptions is consistent.

![A stepper with descriptions for all four steps.](/assets/components/stepper/list/stepper-list-description-do.png)

!!!

!!! Dont

Don't mix and match inclusion of descriptions.

![A stepper where steps one and four have description but steps two and three only have titles.](/assets/components/stepper/list/stepper-list-description-dont.png)

!!!

## Progressive disclosure

> We recommend showing the contents of only the **in-progress** step while hiding the contents of other steps. This narrows the focus to only the current step and ensures steps are completed in order. However, this isn’t always feasible, and sometimes context from the previous step is needed for the user to complete the next step. Use your best judgment.

Reviewer: [nit suggestion] 
```suggestion
We recommend showing the contents of the **in-progress** step while hiding the contents of other steps. This narrows the focus to the current step and ensures steps are completed in order. However, context from the previous step is sometimes needed for the user to complete the next step, so use your best judgment.
```

!!! Do 

> Only show one call to action at a time.

Reviewer: "Call to action" refers to the buttons or links present, not progressive disclosure. What about something like: 

Show the contents of one step at a time.

![Stepper with 3 steps. The current active step is step 1 and it has a call to action button in the content for the user to do the action required for the step.](/assets/components/stepper/list/stepper-list-progressive-disclosure-do.png)

!!!

!!! Dont

> Reduce visual noise by focusing which action the user should be taking first.

Reviewer 4: Do's & Dont's should be formulated to read as:

* "Do this"
* "Don't do this"

to be more easily understood by someone reading through them.

> Reduce visual noise by focusing which action the user should be taking first.

Author: ```suggestion
Be mindful of visual noise. Instead focus which action the user should be taking first.
```

> Reduce visual noise by focusing which action the user should be taking first.

Reviewer 4: ```suggestion
Don’t introduce visual noise. Instead, focus the user’s attention on which action should be taken first.
```

> Reduce visual noise by focusing which action the user should be taking first.

Reviewer 4: Resolved with another applied suggestion
